File: src/lib/tauri/watermark.ts
async function invoke<T>(cmd: string, args?: Record<string, unknown>): Promise<T> {
  const { invoke: tauriInvoke } = await import("@tauri-apps/api/core");
  return tauriInvoke<T>(cmd, args);
}

export interface ApplyWatermarkArgs {
  inputPath: string;
  outputPath: string;
  position: string;
  imagePath?: string;
  text?: string;
  fontPath?: string;
  fontSize?: number;
  videoCodec?: string;
  crf?: number;
  durationSecs?: number;
  jobId?: string;
}

export function applyWatermark(args: ApplyWatermarkArgs): Promise<void> {
  return invoke<void>("apply_watermark", {
    input_path: args.inputPath,
    output_path: args.outputPath,
    position: args.position,
    image_path: args.imagePath ?? null,
    text: args.text ?? null,
    font_path: args.fontPath ?? null,
    font_size: args.fontSize ?? null,
    video_codec: args.videoCodec ?? null,
    crf: args.crf ?? null,
    duration_secs: args.durationSecs ?? null,
    job_id: args.jobId ?? null,
  });
}

File: src-tauri/src/commands/watermark.rs
use tauri::AppHandle;

use crate::models::error::AppError;
use crate::services::watermark::WatermarkService;

/// Burn an image or text overlay onto a video.
///
/// Provide `image_path` for a logo, or `text` for drawtext. Image wins if both
/// are set. `position` is `tl` / `tr` / `bl` / `br` / `center`.
#[tauri::command(rename_all = "snake_case")]
pub async fn apply_watermark(
    app: AppHandle,
    input_path: String,
    output_path: String,
    position: String,
    image_path: Option<String>,
    text: Option<String>,
    font_path: Option<String>,
    font_size: Option<u32>,
    video_codec: Option<String>,
    crf: Option<u8>,
    duration_secs: Option<f64>,
    job_id: Option<String>,
) -> Result<(), AppError> {
    if input_path.is_empty() || output_path.is_empty() {
        return Err(AppError::InvalidArgument(
            "input_path and output_path must not be empty".into(),
        ));
    }
    if let Some(image) = image_path.filter(|s| !s.is_empty()) {
        return WatermarkService::apply_image(
            &app,
            &input_path,
            &image,
            &output_path,
            &position,
            video_codec.as_deref(),
            crf,
            duration_secs,
            job_id.as_deref(),
        )
        .await;
    }
    let text = text.filter(|s| !s.is_empty()).ok_or_else(|| {
        AppError::InvalidArgument("provide image_path or text for the watermark".into())
    })?;
    WatermarkService::apply_text(
        &app,
        &input_path,
        &output_path,
        &text,
        &position,
        font_path.as_deref(),
        font_size.unwrap_or(32),
        video_codec.as_deref(),
        crf,
        duration_secs,
        job_id.as_deref(),
    )
    .await
}
